File: training-data/go/longest_substring_no_repeat.go
package main

import "fmt"

func longestUnique(s string) string {
	last := map[rune]int{}
	runes := []rune(s)
	start, bestStart, bestLen := 0, 0, 0
	for i, r := range runes {
		if p, ok := last[r]; ok && p >= start {
			start = p + 1
		}
		last[r] = i
		if i-start+1 > bestLen {
			bestStart, bestLen = start, i-start+1
		}
	}
	return string(runes[bestStart : bestStart+bestLen])
}

func main() {
	for _, s := range []string{"abcabcbb", "bbbbb", "pwwkew", "dvdf", "héllo wörld"} {
		r := longestUnique(s)
		fmt.Println(s, r, len([]rune(r)))
	}
}

File: training-data/javascript/longest_substring_no_repeat.js
function longestUnique(s) {
  const last = new Map();
  let start = 0, bestStart = 0, bestLen = 0;
  for (let i = 0; i < s.length; i++) {
    if (last.has(s[i]) && last.get(s[i]) >= start) start = last.get(s[i]) + 1;
    last.set(s[i], i);
    if (i + 1 - start > bestLen) { bestLen = i + 1 - start; bestStart = start; }
  }
  return s.slice(bestStart, bestStart + bestLen);
}

for (const s of ["abcabcbb", "bbbbb", "pwwkew", "dvdf"]) {
  const r = longestUnique(s);
  console.log(s, r, r.length);
}
module.exports = { longestUnique };
